Sources/NIOPosix/Pool.swift

```diff
@@ -202,3 +202,117 @@ extension Int {
+
+        var totalByteCount: Int {
+            self.spaceForMsgHdrs + self.spaceForAddresses + MemoryLayout<MemorySentinel>.size
+        }
+
+        var msgHdrsOffset: Int {
+            0
+        }
+
+        var addressesOffset: Int {
+            self.spaceForMsgHdrs
+        }
+
+        var memorySentinelOffset: Int {
+            return self.spaceForMsgHdrs + self.spaceForAddresses
+        }
+    }
+
+    private class BackingStorage: ManagedBuffer<PooledMsgBufferHead, UInt8> {
+        static func create(count: Int) -> Self {
+            let head = PooledMsgBufferHead(count: count)
+
+            let baseStorage = Self.create(minimumCapacity: head.totalByteCount) { _ in
+                head
+            }
+
+            let storage = unsafeDowncast(baseStorage, to: Self.self)
+            storage.withUnsafeMutablePointers { headPointer, tailPointer in
+                UnsafeRawPointer(tailPointer + headPointer.pointee.msgHdrsOffset).bindMemory(to: MMsgHdr.self, capacity: count)
+                UnsafeRawPointer(tailPointer + headPointer.pointee.addressesOffset).bindMemory(to: sockaddr_storage.self, capacity: count)
+                UnsafeRawPointer(tailPointer + headPointer.pointee.memorySentinelOffset).bindMemory(to: MemorySentinel.self, capacity: 1)
+            }
+
+            return storage
+        }
+
+        func withUnsafeMutableTypedPointers<ReturnType>(
+            _ body: (UnsafeMutableBufferPointer<MMsgHdr>, UnsafeMutableBufferPointer<sockaddr_storage>, UnsafeMutablePointer<MemorySentinel>) throws -> ReturnType
+        ) rethrows -> ReturnType {
+            return try self.withUnsafeMutablePointers { headPointer, tailPointer in
+                let msgHdrsPointer = UnsafeMutableRawPointer(tailPointer + headPointer.pointee.msgHdrsOffset).assumingMemoryBound(to: MMsgHdr.self)
+                let addressesPointer = UnsafeMutableRawPointer(tailPointer + headPointer.pointee.addressesOffset).assumingMemoryBound(to: sockaddr_storage.self)
+                let sentinelPointer = UnsafeMutableRawPointer(tailPointer + headPointer.pointee.memorySentinelOffset).assumingMemoryBound(to: MemorySentinel.self)
+
+                let msgHdrsBufferPointer = UnsafeMutableBufferPointer(
+                    start: msgHdrsPointer, count: headPointer.pointee.count
+                )
+                let addressesBufferPointer = UnsafeMutableBufferPointer(
+                    start: addressesPointer, count: headPointer.pointee.count
+                )
+                return try body(msgHdrsBufferPointer, addressesBufferPointer, sentinelPointer)
+            }
+        }
+    }
+
+    private func validateSentinel() {
+        self.storage.withUnsafeMutableTypedPointers { _, _, sentinelPointer in
+            precondition(sentinelPointer.pointee == Self.sentinelValue, "Detected memory handling error!")
+        }
+    }
+
+    private var storage: BackingStorage
+    public var controlMessageStorage: UnsafeControlMessageStorage
```

Comment:
I think this thing is why this is currently a class. Presumably we're going to have to pool this too, so my gut instinct is that we should omit the change to this right now, and have a separate change to clean up the lifecycle management and move this to a pool.

Reply:
Do you mean remove the controlMessageStorage from the PooledMsgBuffer and move it back to the EL for now?

Reply:
Yes.

Reply:
Done.